python/vyos/task_scheduler.py

```diff
@@ -0,0 +1,109 @@
+# Copyright 2020-2022 VyOS maintainers and contributors <maintainers@vyos.io>
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import re
+
+from vyos import ConfigError
+
+def format_task(minute="*", hour="*", day="*", dayofweek="*", month="*", user="root", rawspec=None, command=""):
+    fmt_full = "{minute} {hour} {day} {month} {dayofweek} {user} {command}\n"
+    fmt_raw = "{spec} {user} {command}\n"
+
+    if rawspec is None:
+        s = fmt_full.format(minute=minute, hour=hour, day=day,
+                            dayofweek=dayofweek, month=month, command=command, user=user)
+    else:
+        s = fmt_raw.format(spec=rawspec, user=user, command=command)
+
+    return s
+
+def split_interval(s):
+    result = re.search(r"(\d+)([mdh]?)", s)
+    value = int(result.group(1))
+    suffix = result.group(2)
+    return( (value, suffix) )
+
+def make_command(executable, arguments):
+    if arguments:
+        return("sg vyattacfg \"{0} {1}\"".format(executable, arguments))
+    else:
+        return("sg vyattacfg \"{0}\"".format(executable))
+
+def task_scheduler_verify(tasks):
```

Comment:
Sorry, bit of miscommunication. I didn't mean for you to move all of the task scheduler functions to a module.

I believe it is best to have just the crontab formatting logic from within task_scheduler `generate()`  moved into a module and referencing that code in both `external-lists` and `task-scheduler`.

Reply:
I'm sorry I don't understand, I did only move what was required.

I need verify, generate and apply.  The other parts are dependancies of those 3.  Verify should always be called to make sure there are no config errors before generate, and apply should always be called (in this case it's just a pass) but in the future if there is code there to restart the cron daemon or something it really need to be in the code so that it is future proof.

So I'm sorry but I still do not understand what parts I moved that shouldn't have?  The parts of task-scheduler that are proprietary like the get_config, and the crontab_filename etc etc is all still in the conf_mod task_scheduler

Could you tell me the line numbers maybe that would help me understand?

Reply:
The verify steps shouldn't be needed as the XML constraints should be sufficient to prevent erroneous values.

I believe the only shared code needed would be this block [here](https://github.com/vyos/vyos-1x/pull/1648/files#diff-01e145e011c80f0c91c0f959e077929a8aaeb3267fb8310dd7ba13d09b9bf88dR88), parsing the interval and command into crontab lines.

Reply:
I respectfully disagree, firstly generate called format_task, make_command and split_interval.

So really the only thing being removed would be verify.  And again I don't think that is wise, verify should always be run, honestly verify should be run by generate to make sure the input is safe and correct.  In the future that layout could have breaking changes and if a module is forgotten it could send breaking data to generate.

To be safe and future-roof every module shoudl always rung verify -> generate -> apply, this isn't just for tasks this should be a standard for any type of module that sends in data to be generated for any other daemon or process.

But again we are discussing moving one single function back since generate needs the other 3.  It doesn't seem to make sense to move the safety function but keep everything else.

I would like a second option on this as it just does not make sense to me.

Reply:
A general comment: one wants to maintain the existence (even if, say, verify is a no-op) and structure of the 'canonical functions': get_config, verify, generate, apply. This is the design assumed by authors of conf_mode scripts and by the vyos-configd daemon:
https://docs.vyos.io/en/latest/contributing/development.html#configuration-script-structure-and-behaviour
This structure is checked by the smoketest scripts/cli/test_configd_inspect.py

Reply:
Yes, and that was not meant as a critique of the PR, just a general reminder as you and @sarthurdev debate factorization. Looking at your verify function in task_schedular, it appears to be the right use of that function, and necessary, in order to enforce an exclusive choice. Ignoring that piece for a moment, @sarthurdev 's suggestion of moving the crontab formatting to a utility function makes sense; all the other verify/generate logic can just live in the conf_mode script, as I gather you originally had it.

Reply:
Oh sorry if I sounded offended, I'm not, I really want to work with you guys on things, I'm just at a loss for the request.  I did misunderstand as I thought you were agreeing with @sarthurdev 

And sadly still don't understand, I need verify, generate and apply in my scripts to properly make cronjobs safely, so I can not really move any of them or I can't use them.  The other option is to write my own versions but that seem silly, duplicate code is a major no no.  It would cause desync on how things are done and bloat in general.

I still do not see a way to change any part of this without issues.  This is the sanest way I can see, this way there is little to no duplication of code and it's the most future proof.

Reply:
IMHO the cron stuff should be moved entirely out of this module and `system task-scheduler` should be used. That's is one and only intention.